packages/@lwc/engine-core/src/framework/stylesheet.ts

```diff
@@ -117,24 +126,30 @@ export function getStylesheetsContent(vm: VM, template: Template): string[] {
             stylesheets,
             hostSelector,
             shadowSelector,
-            !syntheticShadow
+            !syntheticShadow,
+            hasAdoptedStyleSheets
         );
     }
 
     return content;
 }
 
-export function createStylesheet(vm: VM, stylesheets: string[]): VNode | null {
-    const { renderer } = vm;
+export function createStylesheet(vm: VM, stylesheets: StylesheetFactoryResult[]): VNode | null {
+    const { renderer, cmpRoot } = vm;
```

Comment:
The fact that we have to add type guards everywhere in this function makes me think that we should attempt to constraint further the types.

Reply:
Not sure I understand what you mean? Note I'm new to TypeScript, so still figuring it out. 🙂

Reply:
The fact that you have to use `as` multiple times here to differentiate between the different kinds of objects the `stylesheets` contains is usually a code smell in TypeScript. Usually, you can enforce such constraint by making the type more "intelligent". 

To be honest with you I don't see how to satisfy this without creating a new abstraction (and impacting performance). Let's leave it as is.